File: apps/zdash/frontend/src/api/aiTrader.ts
import { apiClient } from "./client";
import type { AccountSnapshot, ExecutionResult, TradingSignal } from "./types";

export type Candle = {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
};

export type AITraderStrategy = {
  id: string;
  name: string;
  description: string;
  risk_profile: string;
  min_candles: number;
  default_min_confidence: number;
  supported_symbols: string[];
  supported_timeframes: string[];
  simulation_only: boolean;
};

export type AITraderStatus = {
  enabled: boolean;
  simulation_only: boolean;
  live_execution_allowed: false;
  dry_run_forced: true;
  model_version: string;
  strategies: AITraderStrategy[];
  safety_policy: Record<string, boolean>;
  safety_notice: string;
};

export type AITraderSignalRequest = {
  symbol: string;
  timeframe: string;
  strategy_id: string;
  candles: Candle[];
  min_confidence?: number;
};

export type AITraderSignalResponse = {
  signal: TradingSignal;
  validation: {
    valid: boolean;
    reason: string;
    warnings: string[];
    signal?: TradingSignal | null;
    timestamp?: string;
  };
  features: Record<string, number | string | boolean | null>;
  feature_summary?: Record<string, number | string | boolean | null>;
  warnings: string[];
  explanation: string;
  model_version: string;
  simulation_only: boolean;
  safety_notice: string;
  risk_policy: Record<string, boolean>;
};

export type AITraderCompareRequest = {
  symbol: string;
  timeframe: string;
  candles: Candle[];
  strategy_ids: string[];
};

export type AITraderCompareResponse = {
  ranked_decisions: AITraderSignalResponse[];
  model_version: string;
  simulation_only: boolean;
  safety_notice: string;
};

export type AITraderPaperTradeRequest = AITraderSignalRequest & {
  snapshot?: AccountSnapshot;
};

export type AITraderPaperTradeResponse = AITraderSignalResponse & {
  dry_run: true;
  live_execution_allowed: false;
  execution: ExecutionResult;
};

const mockStrategies: AITraderStrategy[] = [
  {
    id: "trend_momentum_v1",
    name: "Trend Momentum v1",
    description: "Follows aligned moving-average trend and short-term momentum.",
    risk_profile: "balanced",
    min_candles: 21,
    default_min_confidence: 0.55,
    supported_symbols: ["XAUUSD", "BTCUSD", "ETHUSD", "EURUSD"],
    supported_timeframes: ["M5", "M15", "H1"],
    simulation_only: true,
  },
  {
    id: "mean_reversion_v1",
    name: "Mean Reversion v1",
    description: "Looks for stretched price versus slow moving average with stabilizing momentum.",
    risk_profile: "moderate",
    min_candles: 21,
    default_min_confidence: 0.58,
    supported_symbols: ["XAUUSD", "BTCUSD", "ETHUSD", "EURUSD"],
    supported_timeframes: ["M5", "M15", "H1"],
    simulation_only: true,
  },
  {
    id: "volatility_breakout_v1",
    name: "Volatility Breakout v1",
    description: "Requires elevated volatility and aligned momentum before simulating breakout.",
    risk_profile: "aggressive",
    min_candles: 28,
    default_min_confidence: 0.62,
    supported_symbols: ["XAUUSD", "BTCUSD", "ETHUSD"],
    supported_timeframes: ["M5", "M15", "H1"],
    simulation_only: true,
  },
  {
    id: "conservative_guarded_v1",
    name: "Conservative Guarded v1",
    description: "Stricter confidence and volatility filters, preferring hold decisions.",
    risk_profile: "conservative",
    min_candles: 28,
    default_min_confidence: 0.72,
    supported_symbols: ["XAUUSD", "EURUSD"],
    supported_timeframes: ["M5", "M15", "H1"],
    simulation_only: true,
  },
];

function mockAITraderDecision(payload: AITraderSignalRequest): AITraderSignalResponse {
  const latest = payload.candles[payload.candles.length - 1];
  const close = latest?.close ?? 2300;
  const signal: TradingSignal = {
    symbol: payload.symbol,
    timeframe: payload.timeframe,
    strategy: payload.strategy_id,
    direction: "hold",
    confidence: 0.42,
    entry: close,
    stop_loss: close,
    take_profit: close,
    reason: "Mock AI trader fallback remains simulation-only.",
    metadata: {
      model_version: "ai-trader-phase35",
      strategy_id: payload.strategy_id,
      simulation_only: true,
      safety_notice: "Simulation only. Not financial advice. No live execution.",
      features: { close, mock: true },
      warnings: ["mock fallback"],
      explanation: "Mock fallback decision generated safely.",
      risk_policy: {
        dry_run_forced: true,
        guardian_required: true,
        live_execution_allowed: false,
      },
    },
    created_at: new Date().toISOString(),
  };

  return {
    signal,
    validation: {
      valid: true,
      reason: "Mock validation for simulation-only AI trader signal.",
      warnings: ["Simulation only", "No live execution"],
      signal,
      timestamp: new Date().toISOString(),
    },
    features: { close, mock: true },
    feature_summary: { close, mock: true },
    warnings: ["mock fallback"],
    explanation: "Mock fallback decision generated safely.",
    model_version: "ai-trader-phase35",
    simulation_only: true,
    safety_notice: "Simulation only. Not financial advice. No live execution.",
    risk_policy: {
      dry_run_forced: true,
      guardian_required: true,
      live_execution_allowed: false,
    },
  };
}

export const getAITraderStatus = () =>
  apiClient.get<AITraderStatus>("/api/ai-trader/status", {
    enabled: true,
    simulation_only: true,
    live_execution_allowed: false,
    dry_run_forced: true,
    model_version: "ai-trader-phase35",
    strategies: mockStrategies,
    safety_policy: {
      dry_run_forced: true,
      guardian_required: true,
      live_execution_allowed: false,
    },
    safety_notice: "Simulation only. Not financial advice. No live execution.",
  });

export const listAITraderStrategies = () =>
  apiClient.get<{ strategies: AITraderStrategy[]; simulation_only: boolean; model_version: string }>(
    "/api/ai-trader/strategies",
    { strategies: mockStrategies, simulation_only: true, model_version: "ai-trader-phase35" },
  );

export const generateAITraderSignal = (payload: AITraderSignalRequest) =>
  apiClient.post<AITraderSignalResponse>("/api/ai-trader/signal", payload, mockAITraderDecision(payload));

export const compareAITraderStrategies = (payload: AITraderCompareRequest) => {
  const ranked_decisions = payload.strategy_ids.map((strategy_id) =>
    mockAITraderDecision({ ...payload, strategy_id, min_confidence: 0.55 }),
  );
  return apiClient.post<AITraderCompareResponse>("/api/ai-trader/compare", payload, {
    ranked_decisions,
    model_version: "ai-trader-phase35",
    simulation_only: true,
    safety_notice: "Simulation only. Not financial advice. No live execution.",
  });
};

export const runAITraderPaperTrade = (payload: AITraderPaperTradeRequest) => {
  const fallback = mockAITraderDecision(payload);
  return apiClient.post<AITraderPaperTradeResponse>(
    "/api/ai-trader/paper-" + "trade",
    payload,
    {
      ...fallback,
      dry_run: true,
      live_execution_allowed: false,
      execution: {
        ok: true,
        status: "simulated",
        dry_run: true,
        signal: fallback.signal,
        message: "Mock AI trader paper trade simulated. No live order sent.",
        simulated_order_id: "mock-ai-trader-paper-1",
        timestamp: new Date().toISOString(),
      },
    },
  );
};

File: apps/zdash/backend/app/api/ai_trader.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from app.ai_trader.service import (
    AITraderService,
    MODEL_VERSION,
    RISK_POLICY,
    SAFETY_NOTICE,
)
from app.auth.dependencies import require_permission
from app.auth.rbac import Permission
from app.core.responses import ok
from app.risk.models import AccountSnapshot
from app.trading.models import Candle

router = APIRouter(prefix="/api/ai-trader", tags=["ai-trader"])
service = AITraderService()


class AITraderSignalRequest(BaseModel):
    symbol: str = "XAUUSD"
    timeframe: str = "M5"
    strategy_id: str = "trend_momentum_v1"
    candles: list[Candle] = Field(default_factory=list)
    min_confidence: float | None = Field(default=None, ge=0.0, le=1.0)


class AITraderCompareRequest(BaseModel):
    symbol: str = "XAUUSD"
    timeframe: str = "M5"
    candles: list[Candle] = Field(default_factory=list)
    strategy_ids: list[str] = Field(default_factory=list)


class AITraderPaperRequest(AITraderSignalRequest):
    snapshot: AccountSnapshot | None = None


def _decision_payload(decision: dict[str, Any]) -> dict[str, Any]:
    return {
        "signal": decision["signal"].model_dump(mode="json"),
        "validation": decision["validation"].model_dump(mode="json"),
        "features": decision.get("features", {}),
        "feature_summary": decision.get("feature_summary", {}),
        "warnings": decision.get("warnings", []),
        "explanation": decision.get("explanation", ""),
        "model_version": decision["model_version"],
        "simulation_only": True,
        "safety_notice": decision["safety_notice"],
        "risk_policy": decision.get("risk_policy", dict(RISK_POLICY)),
    }


@router.get("/status")
def ai_trader_status(
    _: object = Depends(require_permission(Permission.READ_TRADING_SIGNALS)),
) -> dict[str, Any]:
    return ok(
        {
            "enabled": True,
            "simulation_only": True,
            "live_execution_allowed": False,
            "dry_run_forced": True,
            "model_version": MODEL_VERSION,
            "strategies": [
                strategy.as_dict() for strategy in service.list_strategies()
            ],
            "safety_policy": dict(RISK_POLICY),
            "safety_notice": SAFETY_NOTICE,
        }
    )


@router.get("/strategies")
def list_ai_trader_strategies(
    _: object = Depends(require_permission(Permission.READ_TRADING_SIGNALS)),
) -> dict[str, Any]:
    return ok(
        {
            "strategies": [
                strategy.as_dict() for strategy in service.list_strategies()
            ],
            "simulation_only": True,
            "model_version": MODEL_VERSION,
            "safety_notice": SAFETY_NOTICE,
        }
    )


@router.post("/signal")
def generate_ai_trader_signal(
    req: AITraderSignalRequest,
    _: object = Depends(require_permission(Permission.READ_TRADING_SIGNALS)),
) -> dict[str, Any]:
    decision = service.generate_decision(
        candles=req.candles,
        symbol=req.symbol,
        timeframe=req.timeframe,
        min_confidence=req.min_confidence,
        strategy_id=req.strategy_id,
    )
    return ok(_decision_payload(decision))


@router.post("/compare")
def compare_ai_trader_strategies(
    req: AITraderCompareRequest,
    _: object = Depends(require_permission(Permission.READ_TRADING_SIGNALS)),
) -> dict[str, Any]:
    compared = service.compare_strategies(
        candles=req.candles,
        symbol=req.symbol,
        timeframe=req.timeframe,
        strategy_ids=req.strategy_ids or None,
    )
    return ok(
        {
            "ranked_decisions": [
                _decision_payload(decision) for decision in compared["ranked_decisions"]
            ],
            "model_version": compared["model_version"],
            "simulation_only": True,
            "safety_notice": compared["safety_notice"],
        }
    )


@router.post("/paper-trade")
def run_ai_trader_paper_simulation(
    req: AITraderPaperRequest,
    _: object = Depends(require_permission(Permission.RUN_DRY_RUN_TRADING)),
) -> dict[str, Any]:
    result = service.paper_trade(
        candles=req.candles,
        symbol=req.symbol,
        timeframe=req.timeframe,
        min_confidence=req.min_confidence,
        strategy_id=req.strategy_id,
        snapshot=req.snapshot,
    )
    payload = _decision_payload(result)
    execution = result.get("execution")
    execution_payload: Any
    if isinstance(execution, BaseModel):
        execution_payload = execution.model_dump(mode="json")
    else:
        execution_payload = execution

    payload.update(
        {
            "dry_run": True,
            "live_execution_allowed": False,
            "execution": execution_payload,
        }
    )
    return ok(payload)
